aws/resource_aws_dlm_lifecycle_policy.go

```diff
@@ -0,0 +1,344 @@
+					dlm.SettablePolicyStateValuesDisabled,
+					dlm.SettablePolicyStateValuesEnabled,
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsDlmLifecyclePolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	input := dlm.CreateLifecyclePolicyInput{
+		Description:      aws.String(d.Get("description").(string)),
+		ExecutionRoleArn: aws.String(d.Get("execution_role_arn").(string)),
+		PolicyDetails:    expandDlmPolicyDetails(d.Get("policy_details").([]interface{})),
+		State:            aws.String(d.Get("state").(string)),
+	}
+
+	log.Printf("[INFO] Creating DLM lifecycle policy: %s", input)
+	out, err := conn.CreateLifecyclePolicy(&input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.PolicyId)
+
+	return resourceAwsDlmLifecyclePolicyRead(d, meta)
+}
+
+func resourceAwsDlmLifecyclePolicyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	log.Printf("[INFO] Reading DLM lifecycle policy: %s", d.Id())
+	out, err := conn.GetLifecyclePolicy(&dlm.GetLifecyclePolicyInput{
+		PolicyId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" && !d.IsNewResource() {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("description", out.Policy.Description)
+	d.Set("execution_role_arn", out.Policy.ExecutionRoleArn)
+	d.Set("state", out.Policy.State)
+	if err := d.Set("policy_details", flattenDlmPolicyDetails(out.Policy.PolicyDetails)); err != nil {
+		return fmt.Errorf("error setting policy details %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsDlmLifecyclePolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	input := dlm.UpdateLifecyclePolicyInput{
+		PolicyId: aws.String(d.Id()),
+	}
+
+	if d.HasChange("description") {
+		input.Description = aws.String(d.Get("description").(string))
+	}
+	if d.HasChange("execution_role_arn") {
+		input.ExecutionRoleArn = aws.String(d.Get("execution_role_arn").(string))
+	}
+	if d.HasChange("state") {
+		input.State = aws.String(d.Get("state").(string))
+	}
+	if d.HasChange("policy_details") {
+		input.PolicyDetails = expandDlmPolicyDetails(d.Get("policy_details").([]interface{}))
+	}
+
+	log.Printf("[INFO] Updating lifecycle policy %s", d.Id())
+	_, err := conn.UpdateLifecyclePolicy(&input)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsDlmLifecyclePolicyRead(d, meta)
+}
+
+func resourceAwsDlmLifecyclePolicyDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dlmconn
+
+	log.Printf("[INFO] Deleting DLM lifecycle policy: %s", d.Id())
+	_, err := conn.DeleteLifecyclePolicy(&dlm.DeleteLifecyclePolicyInput{
+		PolicyId: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func expandDlmPolicyDetails(cfg []interface{}) *dlm.PolicyDetails {
+	policyDetails := &dlm.PolicyDetails{}
+	m := cfg[0].(map[string]interface{})
```

Comment:
To prevent potential panics, we should length and nil check the slice before accessing its first element, e.g.

```go
if len(cfg) == 0 || cfg[0] == nil {
  return nil
}
m := cfg[0].(map[string]interface{})
```

Reply:
Ack yeah, I keep missing these. Is there a linting rule or something we can use to check for this?